volplugin/init.go

```diff
@@ -2,67 +2,107 @@ package volplugin
 
 import (
 	"strings"
+	"time"
 
 	"golang.org/x/net/context"
 
 	"github.com/contiv/errored"
 	"github.com/contiv/volplugin/config"
 	"github.com/contiv/volplugin/errors"
 	"github.com/contiv/volplugin/lock"
+	"github.com/contiv/volplugin/storage"
 	"github.com/contiv/volplugin/storage/backend"
 	"github.com/docker/engine-api/client"
 	"github.com/docker/engine-api/types"
 
 	log "github.com/Sirupsen/logrus"
 )
 
-func (dc *DaemonConfig) updateMounts() error {
-	dockerClient, err := client.NewEnvClient()
-	if err != nil {
-		return errored.Errorf("Could not initiate docker client").Combine(err)
-	}
+func (dc *DaemonConfig) getMounted() (map[string]*storage.Mount, map[string]int, error) {
+	mounts := map[string]*storage.Mount{}
+	counts := map[string]int{}
 
-	containers, err := dockerClient.ContainerList(context.Background(), types.ContainerListOptions{})
-	if err != nil {
-		return errored.Errorf("Could not query docker").Combine(err)
-	}
+	now := time.Now()
 
-	for _, container := range containers {
-		if container.State == "running" {
-			for _, mount := range container.Mounts {
-				if mount.Driver != "volplugin" {
-					continue
+	// XXX this loop will indefinitely run if the docker service is down.
+	// This is intentional to ensure we don't take any action when docker is down.
+	for {
+		dockerClient, err := client.NewEnvClient()
+		if err != nil {
+			return nil, nil, errored.Errorf("Could not initiate docker client").Combine(err)
+		}
+
+		containers, err := dockerClient.ContainerList(context.Background(), types.ContainerListOptions{})
+		if err != nil {
+			if now.Sub(time.Now()) > dc.Global.Timeout {
+				panic("Cannot contact docker")
+			}
+			log.Error(errored.Errorf("Could not query docker; retrying").Combine(err))
+			time.Sleep(time.Second)
+			continue
+		}
+
+		for _, container := range containers {
+			if container.State == "running" {
+				for _, mount := range container.Mounts {
+					if mount.Driver == "volplugin" {
+						mounts[mount.Name] = nil
+						counts[mount.Name]++
+					}
 				}
-				dc.mountCounter.Add(mount.Name)
 			}
 		}
+
+		break
 	}
 
 	for driverName := range backend.MountDrivers {
 		cd, err := backend.NewMountDriver(driverName, dc.Global.MountPath)
 		if err != nil {
-			return err
+			return nil, nil, err
 		}
 
-		mounts, err := cd.Mounted(dc.Global.Timeout)
+		mounted, err := cd.Mounted(dc.Global.Timeout)
 		if err != nil {
-			return err
+			return nil, nil, err
+		}
+
+		for _, mount := range mounted {
+			log.Debugf("Refreshing existing mount for %q: %v", mount.Volume.Name, *mount)
+			mounts[mount.Volume.Name] = mount
 		}
+	}
 
-		for _, mount := range mounts {
-			parts := strings.Split(mount.Volume.Name, "/")
+	return mounts, counts, nil
+}
+
+func (dc *DaemonConfig) updateMounts() error {
+	// this loop is used to ensure that volplugin itself has a clear understanding of what's going on.
+	//
+	// FIXME If volplugin detected no mount but docker thinks there is a volplugin
```

Comment:
Just for my (and maybe @yuva29's) edification, what steps can result in volplugin/docker being in this situation?